stable trace state

sorenoid opened the pull request:
<!-- PRs should provide sufficient context on the changes, either by linking to the associated issue and/or by providing a summary. Also provide a link to the design if it's appropriate. -->

Related to issue: https://devtopia.esri.com/runtime/kotlin/issues/4652

<!-- link to design, if applicable -->

### Description:

annotate TraceState as stable
make TraceStaet public/internal property types State<T>
push navController into the TraceNavigation composable because it is an unstable parameter
adjust some navigation logic that used a StateFlow (now uses a State object)


### Pre-merge Checklist

<!-- Tick one box for each section -->
- a [vTest](https://runtime-kotlin.esri.com/view/all/job/vtest/job/toolkit/) Job for this PR has been run
  - [x] link: https://runtime-kotlin.esri.com/view/vTest/job/vtest/job/toolkit/108/
- Unit and/or integration tests have been added to exercise this PR's logic, and the tests are passing:
  - [ ] Yes
  - [ ] No

puneet-pdx commented on toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/ui/AddStartingPointScreen.kt:
```diff
@@ -36,19 +34,16 @@ import com.arcgismaps.toolkit.utilitynetworks.TraceState
-            }
-        }
-    }
+    println("UTILITY NETWORK AddStartingPointScreen")
```
this could be removed

puneet-pdx reviewed (commented):
@sorenoid with these changes I am seeing a performance issue, where on initial load the trace tool has obvious blink effect. Where I think the trace tool is entering the composition and then quickly recomposing.

puneet-pdx commented on toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/TraceState.kt:
```diff
@@ -393,6 +408,15 @@ internal data class StartingPoint(val feature: ArcGISFeature, val utilityElement
+    val graphics: List<Graphic>,
+    val featureResults: List<UtilityElement>,
+    val functionResults: List<UtilityTraceFunctionOutput>,
+    val geometryResults: List<Graphic>
```
`geometryResults` are not needed as they are the same as graphics. I had removed these in my last PR

puneet-pdx commented on toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/TraceState.kt:
```diff
@@ -385,6 +388,18 @@ public sealed class AddStartingPointMode {
+ *
+ * @since 200.6.0
+ */
+internal enum class TraceNavRoute {
```
What is the advantage of having this in `TraceState`? We should just have it in `TraceNaviagation`

sorenoid replied:
Previously we had two sources of truth governing navigation, we had the add starting point mode, and we had this enum. This was leading to complex logic in `TraceNavigation`. Now we just have one source of truth governing navigation. With one source of truth, we need to affect it outside of the composition sometimes, so the nav state is persisted in the state object. For example we need to change the nav route from `TraceState.identifyFeatures`.

puneet-pdx commented on toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/ui/TraceResultScreen.kt:
```diff
@@ -64,7 +64,8 @@ internal fun TraceResultScreen(
     onZoomToResults: () -> Unit,
-    onClearAllResults: () -> Unit
+    onClearAllResults: () -> Unit,
+    onDismiss: () -> Unit
```
`onDismiss` is not used, could be removed ?

puneet-pdx reviewed (commented):
@sorenoid few questions/comments

puneet-pdx commented on toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/Trace.kt:
```diff
@@ -42,39 +42,42 @@ internal const val traceSurfaceContentDescription: String = "trace component sur
 
-    val initializationStatus = traceState.initializationStatus.collectAsStateWithLifecycle()
+    LaunchedEffect(traceState) {
+        if (initializationStatus is InitializationStatus.NotInitialized
```
Do we need this check here ? We are already handling it in [TraceState](https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/blob/feature-branches/utilitynetworktrace/toolkit/utilitynetworks/src/main/java/com/arcgismaps/toolkit/utilitynetworks/TraceState.kt#L121)

sorenoid commented on toolkit/utilitynetworks/build.gradle.kts:
```diff
@@ -82,6 +82,11 @@ android {
     }
 }
 
+composeCompiler {
```
I'll remove it, and can add it back when necessary.

sorenoid commented:
@puneet-pdx I've addressed your comments, making some code changes and continuing conversations in the PR. Can you please re-review or continue the discussion threads?

puneet-pdx reviewed (approved):
@sorenoid looks good 👍🏽
